src/protocol/client.c

```diff
@@ -153,247 +391,234 @@ protocolClientNew(const String *name, const String *service, IoRead *read, IoWri
+    ASSERT(!this->pub.open);
 
-    // Expect data-get state before data end get
-    protocolClientStateExpect(this, protocolClientStateDataGet);
+    protocolClientRequestInternal(this, protocolCommandTypeOpen, param.param);
+    this->pub.open = true;
 
-    MEM_CONTEXT_TEMP_BEGIN()
+    // Set a callback to cleanup the session if it was not already done for async
+    if (!this->async)
     {
-        PackRead *response = pckReadNewIo(this->pub.read);
-        ProtocolMessageType type = (ProtocolMessageType)pckReadU32P(response);
-
-        protocolClientError(this, type, response);
-
-        CHECK(FormatError, type == protocolMessageTypeDataEnd, "expected data end message");
-
-        pckReadEndP(response);
-
-        // Switch state to idle after successful data end get
-        this->state = protocolClientStateIdle;
+        lstAdd(this->client->sessionList, &this);
+        memContextCallbackSet(objMemContext(this), protocolClientSessionFreeResource, this);
     }
-    MEM_CONTEXT_TEMP_END();
 
-    FUNCTION_LOG_RETURN_VOID();
+    FUNCTION_LOG_RETURN(PACK_READ, protocolClientResponseInternal(this));
 }
 
 /**********************************************************************************************************************************/
-FN_EXTERN void
-protocolClientCommandPut(ProtocolClient *const this, ProtocolCommand *const command, const bool dataPut)
+FN_EXTERN PackRead *
+protocolClientSessionRequest(ProtocolClientSession *const this, const ProtocolClientRequestParam param)
 {
     FUNCTION_LOG_BEGIN(logLevelTrace);
-        FUNCTION_LOG_PARAM(PROTOCOL_CLIENT, this);
-        FUNCTION_LOG_PARAM(PROTOCOL_COMMAND, command);
-        FUNCTION_LOG_PARAM(BOOL, dataPut);
+        FUNCTION_LOG_PARAM(PROTOCOL_CLIENT_SESSION, this);
+        FUNCTION_LOG_PARAM(PACK_WRITE, param.param);
     FUNCTION_LOG_END();
 
     ASSERT(this != NULL);
-    ASSERT(command != NULL);
 
-    // Expect idle state before command put
-    protocolClientStateExpect(this, protocolClientStateIdle);
+    protocolClientRequestInternal(this, protocolCommandTypeProcess, param.param);
 
-    // Switch state to cmd-put
-    this->state = protocolClientStateDataPut;
+    FUNCTION_LOG_RETURN(PACK_READ, protocolClientResponseInternal(this));
+}
 
-    // Put command
-    protocolCommandPut(command, this->write);
+/**********************************************************************************************************************************/
+FN_EXTERN void
+protocolClientSessionRequestAsync(ProtocolClientSession *const this, const ProtocolClientRequestParam param)
+{
+    FUNCTION_LOG_BEGIN(logLevelTrace);
+        FUNCTION_LOG_PARAM(PROTOCOL_CLIENT_SESSION, this);
+        FUNCTION_LOG_PARAM(PACK_WRITE, param.param);
+    FUNCTION_LOG_END();
 
-    // Switch state to data-get/data-put after successful command put
-    this->state = dataPut ? protocolClientStateCommandDataGet : protocolClientStateDataGet;
+    ASSERT(this != NULL);
+    ASSERT(this->async);
```

Comment:
If you called this function on a non-async session (without the assert) it would end up being async, i.e. return immediately with no response) which would be a bug since the caller would not be expecting it.